Importing libraries

In [2]:
import pandas as pd
import numpy as np

Importing the dataset

In [3]:
tcc = pd.read_csv("dataset.csv")
tcc.info()

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 33 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CustomerID         7043 non-null   str    
 1   Count              7043 non-null   int64  
 2   Country            7043 non-null   str    
 3   State              7043 non-null   str    
 4   City               7043 non-null   str    
 5   Zip Code           7043 non-null   int64  
 6   Lat Long           7043 non-null   str    
 7   Latitude           7043 non-null   float64
 8   Longitude          7043 non-null   float64
 9   Gender             7043 non-null   str    
 10  Senior Citizen     7043 non-null   str    
 11  Partner            7043 non-null   str    
 12  Dependents         7043 non-null   str    
 13  Tenure Months      7043 non-null   int64  
 14  Phone Service      7043 non-null   str    
 15  Multiple Lines     7043 non-null   str    
 16  Internet Service   7043 non-null   

Data Cleaning

Removing useless labels

In [4]:
tcc=tcc.drop(['CustomerID','Count','Country','State','City','Zip Code','Lat Long','Latitude','Longitude','Churn Label','Churn Score','CLTV','Churn Reason'],axis=1)
tcc.info()

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 20 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Gender             7043 non-null   str    
 1   Senior Citizen     7043 non-null   str    
 2   Partner            7043 non-null   str    
 3   Dependents         7043 non-null   str    
 4   Tenure Months      7043 non-null   int64  
 5   Phone Service      7043 non-null   str    
 6   Multiple Lines     7043 non-null   str    
 7   Internet Service   7043 non-null   str    
 8   Online Security    7043 non-null   str    
 9   Online Backup      7043 non-null   str    
 10  Device Protection  7043 non-null   str    
 11  Tech Support       7043 non-null   str    
 12  Streaming TV       7043 non-null   str    
 13  Streaming Movies   7043 non-null   str    
 14  Contract           7043 non-null   str    
 15  Paperless Billing  7043 non-null   str    
 16  Payment Method     7043 non-null   

Converting string datatype to numeric

In [5]:
tcc['Total Charges']=pd.to_numeric(tcc['Total Charges'], errors='coerce')
tcc['Total Charges'].isna().sum()  
tcc['Total Charges'] = tcc['Total Charges'].fillna(0)
tcc.info()

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 20 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Gender             7043 non-null   str    
 1   Senior Citizen     7043 non-null   str    
 2   Partner            7043 non-null   str    
 3   Dependents         7043 non-null   str    
 4   Tenure Months      7043 non-null   int64  
 5   Phone Service      7043 non-null   str    
 6   Multiple Lines     7043 non-null   str    
 7   Internet Service   7043 non-null   str    
 8   Online Security    7043 non-null   str    
 9   Online Backup      7043 non-null   str    
 10  Device Protection  7043 non-null   str    
 11  Tech Support       7043 non-null   str    
 12  Streaming TV       7043 non-null   str    
 13  Streaming Movies   7043 non-null   str    
 14  Contract           7043 non-null   str    
 15  Paperless Billing  7043 non-null   str    
 16  Payment Method     7043 non-null   

Encoding categorical labels

In [6]:
binary_cols = ['Gender', 'Senior Citizen', 'Partner', 'Dependents', 'Phone Service', 'Paperless Billing']
for col in binary_cols:
    tcc[col]=tcc[col].map({'Yes':1,'No':0,
                           'Male':1,'Female':0})
tcc.info()

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 20 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Gender             7043 non-null   int64  
 1   Senior Citizen     7043 non-null   int64  
 2   Partner            7043 non-null   int64  
 3   Dependents         7043 non-null   int64  
 4   Tenure Months      7043 non-null   int64  
 5   Phone Service      7043 non-null   int64  
 6   Multiple Lines     7043 non-null   str    
 7   Internet Service   7043 non-null   str    
 8   Online Security    7043 non-null   str    
 9   Online Backup      7043 non-null   str    
 10  Device Protection  7043 non-null   str    
 11  Tech Support       7043 non-null   str    
 12  Streaming TV       7043 non-null   str    
 13  Streaming Movies   7043 non-null   str    
 14  Contract           7043 non-null   str    
 15  Paperless Billing  7043 non-null   int64  
 16  Payment Method     7043 non-null   

Dropping duplicate values

In [7]:
tcc = tcc.drop_duplicates()
tcc.info()

<class 'pandas.DataFrame'>
Index: 7021 entries, 0 to 7042
Data columns (total 20 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Gender             7021 non-null   int64  
 1   Senior Citizen     7021 non-null   int64  
 2   Partner            7021 non-null   int64  
 3   Dependents         7021 non-null   int64  
 4   Tenure Months      7021 non-null   int64  
 5   Phone Service      7021 non-null   int64  
 6   Multiple Lines     7021 non-null   str    
 7   Internet Service   7021 non-null   str    
 8   Online Security    7021 non-null   str    
 9   Online Backup      7021 non-null   str    
 10  Device Protection  7021 non-null   str    
 11  Tech Support       7021 non-null   str    
 12  Streaming TV       7021 non-null   str    
 13  Streaming Movies   7021 non-null   str    
 14  Contract           7021 non-null   str    
 15  Paperless Billing  7021 non-null   int64  
 16  Payment Method     7021 non-null   str  

Encoding remaining categorical columns

In [8]:
categorical_cols = tcc.select_dtypes(include="object").columns

tcc = pd.get_dummies(tcc, columns=categorical_cols, drop_first=True, dtype=int)
tcc.info()

<class 'pandas.DataFrame'>
Index: 7021 entries, 0 to 7042
Data columns (total 31 columns):
 #   Column                                  Non-Null Count  Dtype  
---  ------                                  --------------  -----  
 0   Gender                                  7021 non-null   int64  
 1   Senior Citizen                          7021 non-null   int64  
 2   Partner                                 7021 non-null   int64  
 3   Dependents                              7021 non-null   int64  
 4   Tenure Months                           7021 non-null   int64  
 5   Phone Service                           7021 non-null   int64  
 6   Paperless Billing                       7021 non-null   int64  
 7   Monthly Charges                         7021 non-null   float64
 8   Total Charges                           7021 non-null   float64
 9   Churn Value                             7021 non-null   int64  
 10  Multiple Lines_No phone service         7021 non-null   int64  
 11  Multipl

C:\Users\csing\AppData\Local\Temp\ipykernel_20560\3997938971.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_cols = tcc.select_dtypes(include="object").columns


Separating Churn Value from dataset as it is the label we have to predict

In [9]:
X = tcc.drop("Churn Value",axis=1)
y=tcc["Churn Value"]
print(X.shape)
print(y.shape)

(7021, 30)
(7021,)


Train/Test Split

In [10]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2, random_state=42,stratify=y)

Standardizing numerical features

In [11]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

numerical_cols = ["Tenure Months", "Monthly Charges", "Total Charges"]

X_train[numerical_cols] = scaler.fit_transform(X_train[numerical_cols])
X_test[numerical_cols] = scaler.transform(X_test[numerical_cols])

Training ANN Model 1
With 300 neurons in first hidden layer
100 neurons in second hidden layer

In [12]:
from sklearn.neural_network import MLPClassifier

mlp_clf = MLPClassifier(
    hidden_layer_sizes=[300, 100],
    verbose=True,
    early_stopping=True,
    random_state=42
)

mlp_clf.fit(X_train, y_train)

y_pred_ann = mlp_clf.predict(X_test)

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    confusion_matrix
)

accuracy = accuracy_score(y_test, y_pred_ann)
precision = precision_score(y_test, y_pred_ann)
recall = recall_score(y_test, y_pred_ann)
f1 = f1_score(y_test, y_pred_ann)

print("ANN Performance")
print("----------------")
print("Accuracy :", accuracy)
print("Precision:", precision)
print("Recall   :", recall)
print("F1-score :", f1)

cm = confusion_matrix(y_test, y_pred_ann)

print("Confusion Matrix:")
print(cm)


Iteration 1, loss = 0.46159317
Validation score: 0.809609
Iteration 2, loss = 0.40625786
Validation score: 0.809609
Iteration 3, loss = 0.39528442
Validation score: 0.807829
Iteration 4, loss = 0.39175231
Validation score: 0.813167
Iteration 5, loss = 0.38696990
Validation score: 0.804270
Iteration 6, loss = 0.38298584
Validation score: 0.791815
Iteration 7, loss = 0.38332618
Validation score: 0.809609
Iteration 8, loss = 0.37638620
Validation score: 0.814947
Iteration 9, loss = 0.37673954
Validation score: 0.816726
Iteration 10, loss = 0.36929332
Validation score: 0.802491
Iteration 11, loss = 0.36713712
Validation score: 0.811388
Iteration 12, loss = 0.36191471
Validation score: 0.795374
Iteration 13, loss = 0.35899398
Validation score: 0.806050
Iteration 14, loss = 0.35714198
Validation score: 0.806050
Iteration 15, loss = 0.35423294
Validation score: 0.813167
Iteration 16, loss = 0.35068956
Validation score: 0.804270
Iteration 17, loss = 0.34454256
Validation score: 0.804270
Iterat

Training ANN Model 2
With 128 neurons in first hidden layer
64 neurons in second hidden layer
And batch size 32

In [13]:
mlp_clf2 = MLPClassifier(
    hidden_layer_sizes=[128, 64],
    learning_rate_init=0.001,
    batch_size=32,
    max_iter=200,
    verbose=True,
    early_stopping=True,
    random_state=42
)

mlp_clf2.fit(X_train, y_train)

y_pred_ann2 = mlp_clf2.predict(X_test)

print("Accuracy :", accuracy_score(y_test, y_pred_ann2))
print("Precision:", precision_score(y_test, y_pred_ann2))
print("Recall   :", recall_score(y_test, y_pred_ann2))
print("F1-score :", f1_score(y_test, y_pred_ann2))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred_ann2))

Iteration 1, loss = 0.45077796
Validation score: 0.820285
Iteration 2, loss = 0.40394442
Validation score: 0.816726
Iteration 3, loss = 0.39840285
Validation score: 0.816726
Iteration 4, loss = 0.39149387
Validation score: 0.818505
Iteration 5, loss = 0.38745909
Validation score: 0.836299
Iteration 6, loss = 0.38523840
Validation score: 0.820285
Iteration 7, loss = 0.38073171
Validation score: 0.836299
Iteration 8, loss = 0.37697165
Validation score: 0.823843
Iteration 9, loss = 0.37256801
Validation score: 0.827402
Iteration 10, loss = 0.36829089
Validation score: 0.830961
Iteration 11, loss = 0.36311151
Validation score: 0.834520
Iteration 12, loss = 0.36132870
Validation score: 0.825623
Iteration 13, loss = 0.35541828
Validation score: 0.832740
Iteration 14, loss = 0.35423176
Validation score: 0.818505
Iteration 15, loss = 0.34803702
Validation score: 0.829181
Iteration 16, loss = 0.34205644
Validation score: 0.802491
Validation score did not improve more than tol=0.000100 for 10 co

Training ANN Model 3
With 128 neurons in first hidden layer
64 neurons in second hidden layer
And batch size 64

In [14]:
mlp_clf3 = MLPClassifier(
    hidden_layer_sizes=[128, 64],
    learning_rate_init=0.0005,
    batch_size=64,
    max_iter=200,
    verbose=True,
    early_stopping=True,
    random_state=42
)

mlp_clf3.fit(X_train, y_train)

y_pred_ann3 = mlp_clf3.predict(X_test)

print("Accuracy :", accuracy_score(y_test, y_pred_ann3))
print("Precision:", precision_score(y_test, y_pred_ann3))
print("Recall   :", recall_score(y_test, y_pred_ann3))
print("F1-score :", f1_score(y_test, y_pred_ann3))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred_ann3))

Iteration 1, loss = 0.51987248
Validation score: 0.802491
Iteration 2, loss = 0.41934533
Validation score: 0.823843
Iteration 3, loss = 0.40663572
Validation score: 0.823843
Iteration 4, loss = 0.39903745
Validation score: 0.818505
Iteration 5, loss = 0.39517963
Validation score: 0.822064
Iteration 6, loss = 0.39240972
Validation score: 0.820285
Iteration 7, loss = 0.38951241
Validation score: 0.818505
Iteration 8, loss = 0.38685975
Validation score: 0.827402
Iteration 9, loss = 0.38687937
Validation score: 0.822064
Iteration 10, loss = 0.38450239
Validation score: 0.823843
Iteration 11, loss = 0.38121456
Validation score: 0.823843
Iteration 12, loss = 0.37920560
Validation score: 0.822064
Iteration 13, loss = 0.37843055
Validation score: 0.830961
Iteration 14, loss = 0.37497322
Validation score: 0.822064
Iteration 15, loss = 0.37271582
Validation score: 0.830961
Iteration 16, loss = 0.37044216
Validation score: 0.820285
Iteration 17, loss = 0.36999287
Validation score: 0.825623
Iterat

Upon comparision, the third model is most suitable for ANN because:
1) Higher F1 Score
2) Better Positive Class Detection

Training Logistic Regression Model

In [15]:
from sklearn.linear_model import LogisticRegression

log_reg = LogisticRegression(
    max_iter=5000,
    random_state=42
)

log_reg.fit(X_train, y_train)

y_pred_lr = log_reg.predict(X_test)

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    confusion_matrix
)

print("Logistic Regression")
print("-------------------")
print("Accuracy :", accuracy_score(y_test, y_pred_lr))
print("Precision:", precision_score(y_test, y_pred_lr))
print("Recall   :", recall_score(y_test, y_pred_lr))
print("F1-score :", f1_score(y_test, y_pred_lr))

print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred_lr))

Logistic Regression
-------------------
Accuracy : 0.805693950177936
Precision: 0.6633663366336634
Recall   : 0.5403225806451613
F1-score : 0.5955555555555555
Confusion Matrix:
[[931 102]
 [171 201]]


Training Random Forest Model

In [16]:
from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier(
    n_estimators=100,
    random_state=42
)

rf.fit(X_train, y_train)

y_pred_rf = rf.predict(X_test)

print("Random Forest")
print("-------------")
print("Accuracy :", accuracy_score(y_test, y_pred_rf))
print("Precision:", precision_score(y_test, y_pred_rf))
print("Recall   :", recall_score(y_test, y_pred_rf))
print("F1-score :", f1_score(y_test, y_pred_rf))

print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred_rf))


Random Forest
-------------
Accuracy : 0.7836298932384341
Precision: 0.6188811188811189
Recall   : 0.47580645161290325
F1-score : 0.5379939209726444
Confusion Matrix:
[[924 109]
 [195 177]]


Final Comparision

In [ ]:
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    confusion_matrix
)
import pandas as pd

# Predictions
y_pred_ann = mlp_clf3.predict(X_test)
y_pred_lr = log_reg.predict(X_test)
y_pred_rf = rf.predict(X_test)


# Calculate metrics
results = {
    "ANN (Experiment 3)": [
        accuracy_score(y_test, y_pred_ann),
        precision_score(y_test, y_pred_ann),
        recall_score(y_test, y_pred_ann),
        f1_score(y_test, y_pred_ann)
    ],

    "Logistic Regression": [
        accuracy_score(y_test, y_pred_lr),
        precision_score(y_test, y_pred_lr),
        recall_score(y_test, y_pred_lr),
        f1_score(y_test, y_pred_lr)
    ],

    "Random Forest": [
        accuracy_score(y_test, y_pred_rf),
        precision_score(y_test, y_pred_rf),
        recall_score(y_test, y_pred_rf),
        f1_score(y_test, y_pred_rf)
    ]
}


# Create comparison table
comparison = pd.DataFrame(
    results,
    index=["Accuracy", "Precision", "Recall", "F1-score"]
)

# Convert to percentage
comparison = comparison * 100

print("Final Model Comparison")
print(comparison.round(2))

# Confusion matrices
print("\nANN (Experiment 3) Confusion Matrix:")
print(confusion_matrix(y_test, y_pred_ann))

print("\nLogistic Regression Confusion Matrix:")
print(confusion_matrix(y_test, y_pred_lr))

print("\nRandom Forest Confusion Matrix:")
print(confusion_matrix(y_test, y_pred_rf))

Final Model Comparison
           ANN (Experiment 3)  Logistic Regression  Random Forest
Accuracy                80.21                80.57          78.36
Precision               67.54                66.34          61.89
Recall                  48.66                54.03          47.58
F1-score                56.56                59.56          53.80

ANN (Experiment 3) Confusion Matrix:
[[946  87]
 [191 181]]

Logistic Regression Confusion Matrix:
[[931 102]
 [171 201]]

Random Forest Confusion Matrix:
[[924 109]
 [195 177]]


Analysis
Answer the following:
1.	Which model provides the best overall performance?
Logistic Regression provides the best overall performance. It has the highest Accuracy (80.57%), Precision (66.34%), Recall (54.03%), and F1-score (59.56%) among the three models.

2.	Which model provides the highest Recall?
Logistic Regression provides the highest Recall, at 54.03%

3.	Why is Recall particularly important for customer churn prediction?
Recall is particularly important in customer churn prediction because it measures how many actual customers who are likely to churn are correctly identified. A low recall means many potential churners are missed, which can result in lost customers. High recall allows the company to identify more at-risk customers and take preventive action.

4.	Which customer characteristics appear to be associated with higher churn?
Characteristics associated with higher churn include shorter tenure, higher monthly charges, month-to-month contracts, electronic check payment, and fiber optic internet service. Customers with fewer long-term commitments also tend to be more likely to churn.

5.	What practical action could a telecom company take using the prediction results?
The telecom company can use the model to identify customers with a high probability of churn and target them with retention strategies, such as discounts, customized plans, loyalty benefits, improved support, or contract incentives. This can help reduce customer loss and improve retention.
